Add docstring for instrument_connection() and tests for database drivers

beijiez opened the pull request:
# Description

Add docstring for `instrument_connection()` and tests for database drivers as it is missing. This will be useful to new database instrumentation efforts

DB driver docstrings updated:

[x] [aiopg](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-aiopg/src/opentelemetry/instrumentation/aiopg/__init__.py)
[x] [mysql-connector](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-mysql/src/opentelemetry/instrumentation/mysql/__init__.py)
[x] [mysqlclient](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-mysqlclient/src/opentelemetry/instrumentation/mysqlclient/__init__.py)
[x] [psycopg](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-psycopg/src/opentelemetry/instrumentation/psycopg/__init__.py)
[x] [psycopg2](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-psycopg2/src/opentelemetry/instrumentation/psycopg2/__init__.py)
[x] [pymysql](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pymysql/src/opentelemetry/instrumentation/pymysql/__init__.py)
[x] [sqlite3](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py)

Addresses #3089 

# How Has This Been Tested?

- Added tests for DB driver `/tests` if it was missing
- Tested Usage code locally
- `tox -e spellcheck` OK
- `tox` congratulations :)

# Does This PR Require a Core Repo Change?
No

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-aiopg/src/opentelemetry/instrumentation/aiopg/__init__.py:
```diff
@@ -35,12 +35,19 @@
     cnx.close()
-
+    
+    cnx = AiopgInstrumentor().instrument_connection(cnx)
```
For this and the other helpful `Usage` examples you've added, please could we also add something so that readers know they don't have to do both `instrument` and `instrument_connection` calls?

It could be like this:

```suggestion
    # Alternatively, use instrument_connection for an individual connection
    cnx = AiopgInstrumentor().instrument_connection(cnx)
```

Or, it could be two separate `.. code-block:: python`. The first block would be for `# Call instrument() to wrap all database connections` and the second block would be for `# Alternatively, use instrument_connection for an individual connection`

beijiez replied:
Makes sense. Updated docstring to have two separate `. code-block:: python` blocks

tammy-baylis-swi replied:
Looks good! Please could you make the same update for mysql, mysqlclient, psycopg, psycopg2, pymysql, sqlite3

tammy-baylis-swi commented:
Thanks @beijiez ! I've requested one set of changes first.

There are a few CI formatting checks failing. Here are some more individual local commands if you could please run on your local:

1. `tox -e ruff`
2. `tox -e lint-instrumentation-psycopg2`
3. `tox -e lint-instrumentation-sqlite3`
4. `tox -e docs` <-- this might only work for Python 3.11 or 3.10

beijiez commented:
> Thanks @beijiez ! I've requested one set of changes first.
> 
> There are a few CI formatting checks failing. Here are some more individual local commands if you could please run on your local:
> 
> 1. `tox -e ruff`
> 2. `tox -e lint-instrumentation-psycopg2`
> 3. `tox -e lint-instrumentation-sqlite3`
> 4. `tox -e docs` <-- this might only work for Python 3.11 or 3.10

Thank you for the fast review @tammy-baylis-swi! Updated and re-ran CI
```
tox -e ruff
tox -e lint-instrumentation-psycopg2
tox -e lint-instrumentation-sqlite3
tox -e lint-instrumentation-aiopg
tox -e docs [on a 3.11.0 pyvenv] 
```

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-mysql/src/opentelemetry/instrumentation/mysql/__init__.py:
```diff
@@ -86,12 +100,16 @@ def instrument_connection(self, connection, tracer_provider=None):
 
         Returns:
-            An instrumented connection.
+            An instrumented MySQL connection with OpenTelemetry tracing enabled,
```
Nit-pick:

```suggestion
            An instrumented MySQL connection with OpenTelemetry tracing enabled.
```

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-mysqlclient/src/opentelemetry/instrumentation/mysqlclient/__init__.py:
```diff
@@ -162,12 +184,28 @@ def instrument_connection(
+                SQL queries will be enriched with contextual information (e.g., database client details).
+                Default is `None`.
+            commenter_options:
+                A dictionary of configuration options for SQLCommenter. This allows you to customize
+                metadata appended to queries. Possible options include:
```
```suggestion
                A dictionary of configuration options for SQLCommenter. This allows you to customize
                metadata appended to queries. All options are enabled (True) by default. Possible options include:
```

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-psycopg/src/opentelemetry/instrumentation/psycopg/__init__.py:
```diff
@@ -88,15 +88,29 @@
+
+    # Alternatively, use instrument_connection for an individual connection
+    cnx = psycopg.connect(database='Database')
+    instrumented_cnx = instrument_connection(cnx)
```
```suggestion
    instrumented_cnx = PsycopgInstrumentor().instrument_connection(cnx)
```

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:
```diff
@@ -37,6 +37,21 @@
+
+    # Alternatively, use instrument_connection for an individual connection
+    conn = sqlite3.connect(":memory:")
+    instrumented_connection = SQLite3Instrumentor.instrument_connection(conn)
```
```suggestion
    instrumented_connection = SQLite3Instrumentor().instrument_connection(conn)
```

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:
```diff
@@ -104,9 +119,14 @@ def instrument_connection(
+            An instrumented SQLite connection that supports
+            telemetry for tracing database operations.
 
+        Notes:
+            - Instrumentation must be explicitly applied to the connection object
+              for tracing to work. This is not done automatically by simply calling
+              `SQLite3Instrumentor().instrument()`.
+        """
```
It should actually work if user does `SQLite3Instrumentor().instrument()` then do e.g. `cnx = sqlite3.connect(':memory:')` and `cursor = cnx.cursor()`. Were there issues when you did the tests update?

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-psycopg2/src/opentelemetry/instrumentation/psycopg2/__init__.py:
```diff
@@ -88,15 +88,29 @@
+
+    # Alternatively, use instrument_connection for an individual connection
+    cnx = psycopg2.connect(database='Database')
+    instrumented_cnx = Psycopg2Instrumentor.instrument_connection(cnx)
```
```suggestion
    instrumented_cnx = Psycopg2Instrumentor().instrument_connection(cnx)
```

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-mysqlclient/src/opentelemetry/instrumentation/mysqlclient/__init__.py:
```diff
@@ -56,6 +56,28 @@
+
+    # Alternatively, use instrument_connection for an individual connection
+    cnx = MySQLdb.connect(database="MySQL_Database")
+    instrumented_cnx = MySQLClientInstrumentor.instrument_connection(
```
```suggestion
    instrumented_cnx = MySQLClientInstrumentor().instrument_connection(
```

tammy-baylis-swi commented:
Thank you again @beijiez ! Nice to see all CI passing now. I've left more suggestions and a question when you have a moment.

beijiez commented:
> Thank you again @beijiez ! Nice to see all CI passing now. I've left more suggestions and a question when you have a moment.

Thank you for the detailed review @tammy-baylis-swi ! I responded to your question, re-verified CI, and pushed changes

Kludex reviewed (changes requested):
All the examples from the code-blocks don't run as they are written. Can we add the table creation before the insertion?

tammy-baylis-swi commented:
> All the examples from the code-blocks don't run as they are written. Can we add the table creation before the insertion?

@Kludex could you provide a `suggestion` on one of the instrumentor scripts for how to do this?

beijiez commented:
> All the examples from the code-blocks don't run as they are written. Can we add the table creation before the insertion?
> 
> It saves time for those reading the docs when trying the code snippets.

Good callout, sqlite3 currently has a working example with table creation. Can add it to rest of the docstrings

```
cursor.execute("CREATE TABLE test (testField INTEGER)")

```

tammy-baylis-swi commented:
> > All the examples from the code-blocks don't run as they are written. Can we add the table creation before the insertion?
> > It saves time for those reading the docs when trying the code snippets.
> 
> Good callout, sqlite3 currently has a working example with table creation. Can add it to rest of the docstrings
> 
> ```
> cursor.execute("CREATE TABLE test (testField INTEGER)")
> ```

Ah, I see. I was thinking docs/formatting tables instead of db tables. 😅

Kludex commented:
Yeah, I updated the sqlite3 one last week. 😁

That would be great. Thanks!

beijiez commented:
> Yeah, I updated the sqlite3 one last week. 😁
> 
> That would be great. Thanks!

Done. Added `cursor.execute("CREATE TABLE IF NOT EXISTS test (testField INTEGER)")` to each to allow idempotent table creation in the example docstrings

tammy-baylis-swi reviewed (approved):
Lgtm, thank you for all the work!

This will be checked by the Maintainers next.
